lib/Sema/ConstraintSystem.cpp

```diff
@@ -2880,6 +2880,58 @@ static bool diagnoseConflictingGenericArguments(ConstraintSystem &cs,
   return diagnosed;
 }
 
+/// Diagnose ambiguity related to overloaded declarations where only
+/// *some* of the overload choices have ephemeral pointer warnings/errors
+/// associated with them. Such situations have be handled specifically
+/// because ephemeral fixes do not affect the score.
+///
+/// If all of the overloads have ephemeral fixes associated with them
+/// it's much easier to diagnose through notes associated with each fix.
+static bool
+diagnoseAmbiguityWithEphemeralPointers(ConstraintSystem &cs,
+                                       ArrayRef<Solution> solutions) {
+  llvm::MapVector<Expr *, ValueDecl *> ambiguities;
+
+  bool allSolutionsHaveFixes = true;
+  for (const auto &solution : solutions) {
+    if (solution.Fixes.empty()) {
+      allSolutionsHaveFixes = false;
+      continue;
+    }
+
+    for (const auto *fix : solution.Fixes) {
+      if (fix->getKind() != FixKind::TreatEphemeralAsNonEphemeral)
+        return false;
+
+      // Ephemeral pointers are only possible in argument positions.
+      if (auto *anchor = simplifyLocatorToAnchor(fix->getLocator())) {
+        anchor = cast<InOutExpr>(anchor)->getSubExpr();
+
+        auto overload = solution.getOverloadChoiceIfAvailable(
+            cs.getConstraintLocator(anchor));
```

Comment:
What if the ambiguity is for the function rather than the argument? For example, something like:

```swift
func foo(@_nonEphemeral _ ptr: UnsafePointer<Int>) {}
func foo(_ ptr: UnsafeRawPointer) {}

func bar(_ x: inout Int) {
  foo(&x)
}
```

I think with this change we'll complain about `x` being ambiguous instead of `foo` being ambiguous.

I wonder if we can defer to `diagnoseAmbiguity` for this? In fact I think we might be able to change the `hasFixes` predicate here:

https://github.com/apple/swift/blob/352653910992aca045782389025c97fb82295aaf/lib/Sema/ConstraintSystem.cpp#L2616

such that rather than filtering out solutions with fixes, it filters out solutions with a non-zero fix score. This should allow us to fall through to `diagnoseAmbiguity` for both this case and more generally for solutions with only warning fixes.

What do you think?

Reply:
I was thinking about that before but I just don't want to add ad hoc checks in `salvage` and `diagnoseAmbiguityWithFixes` to find whether fix is ` TreatEphemeralAsNonEphemeral` so I have decided to do it all in one place, these are fixes after all... For situations you have described I believe argument is going to be the same declaration in all solutions which is a good indication that something else is ambiguous, I'm going add such example into the suite.